### ISSUE TITLE
Can RVO eliminate our return-by-reference overloads?

### ISSUE THREAD

**Issue by nilsvu** [sxs-collaboration/spectre#4648]
Writing a return-by-value and return-by-reference overload adds a considerable amount of boilerplate code, also to resolve the overloads in compute tags or pybindings (see #4637). Return-by-reference overloads are somewhat controversial in the community, with some arguing for letting the compiler optimize away the memory allocations. My question is: does that work reliably with C++17, so we can indeed eliminate the return-by-ref overloads?

The alternative to eliminating the return-by-ref overload is to eliminate the return-by-value overload, which we usually add for convenience. @nilsdeppe and I discussed adding a macro or something that calls the return-by-ref function and returns the result. Any better ideas?

**Comment by wthrowe** [comment]
Isn't RVO only for eliding constructors?  How would this let us use a preexisting allocation for the result?

**Comment by nilsvu** [comment]
You're right. Then we're either stuck with writing a ret-by-val convenience overload or a macro.
